### PR TITLE
React - Controlled Inputs #40535

### PR DESCRIPTION
**Pull request by shashankbaranwal**
Task:
1. Update TextField input to accept onChange handler.
2. TextField must include following props:
* error (default blank, and must only accept string)
* value (string and must be required)
* onChange (function and must be required)
3, Create two more components;:
* SelectField
* RadioGroup
4. Both SelectField and RadioGroup must accept following props:
* error (default blank, and must only accept string)
* value (string and must be required)
* onChange (function and must be required)
* options: (default [], and must contains array of objects with keys *label* and *value*
5. SelectField will accept additional props:
* defaultText: (default as select, and must accept string) - This will be used to show default selected option in select box with value as blank.
6. Now create a new page component name as “InputDemo” similar to “TextFieldDemo”.
7. Replace the TextFieldDemo component with InputDemo component in “App.js”.
8. Now using the above three “Field” components create the layout as shown in attached GIF.
9. Show the state of InputDemo component at every change.
10. Create the handlers for all the fields in input demo and name must start with handle e.g. handleNameChange, handleSportChange and so on.
11. Please note, any of the Input Components must not have their own state.
12.*Don't hardcode any string for comparison like 'football', 'cricket', always use the constant for the same, and add them to configs/constants file.*
13. Options to be used as RadioGroup or Drop Down must be exported from the constants file.
14. Kindly adhere the guidelines of airbnb..

### CODE REVIEW

File: src/config/constants.js [hunk 1]
```diff
@@ -1,7 +1,46 @@
 export const PUBLIC_IMAGE_FOLDER = '/images/';
+export const DEFAULT_BANNER_IMAGE = '/images/banners/default.png';
+export const banners = ['banners/cloud.jpg', 'banners/dns-server.png', 'banners/js.jpg', 'banners/full-stack-web-development.jpg', 'banners/load-balancer.png'];
+export const total = banners.length;
 
-export const DEFAULT_BANNER_IMAGE = 'banners/default.png';
+const selectOptions = [
```
Comment (namanpar) [inline src/config/constants.js:6]: rename it to sports

File: src/config/constants.js [hunk 2]
```diff
@@ -1,7 +1,46 @@
...
+  {
+    label: 'Cricket',
+    value: 'cricket',
+  },
+  {
+    label: 'Football',
+    value: 'football',
+  },
+];
+const radioOptionsCricket = [
```
Comment (namanpar) [inline src/config/constants.js:16]: cricketOptions is good enough. No need to mention select or radio in naming

File: src/config/constants.js [hunk 3]
```diff
@@ -1,7 +1,46 @@
...
+    value: 'bowler',
+  },
+  {
+    label: 'All Rounder',
+    value: 'all rounder',
+  },
+];
 
-export const banners = ['banners/cloud.jpg', 'banners/dns-server.png', 'banners/full-stack-web-development.jpg', 'banners/load-balancer.png'];
+const radioOptionsFootball = [
```
Comment (namanpar) [inline src/config/constants.js:35]: rename it to footballOptions

File: src/pages/InputDemo/InputDemo.jsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                        <RadioField
+                          error=""
+                          options={this.RadioOption()}
+                          onChange={this.handlePositionChange}
+                        />
+                      </>
+                    )
+                }
+              </div>
+            </div>
```
Comment (namanpar) [inline src/pages/InputDemo/InputDemo.jsx:76]: create a function which will return selection option according to input

File: src/pages/InputDemo/InputDemo.jsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        let { radioValue } = this.state;
+        const { sport } = this.state;
+        if (sport === 'cricket') {
+          radioValue = radioOptionsCricket;
+        } else if (sport === 'football') {
+          radioValue = radioOptionsFootball;
+        }
+        return (radioValue);
+      };
+
```
Comment (namanpar) [inline src/pages/InputDemo/InputDemo.jsx:46]: never name function like this. optimize code and remove if-else.

File: src/pages/InputDemo/InputDemo.jsx [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      });
+    }
+
+    handleSportChange = (e) => {
+      this.setState({ sport: e.target.value }, () => console.log(this.state));
+      if (e.target.value === 'Select') {
+        this.setState({ sport: '' });
+      }
+      return e.target.value === 'cricket' ? this.setState({ football: '' }) : this.setState({ cricket: '' });
+    }
```
Comment (namanpar) [inline src/pages/InputDemo/InputDemo.jsx:29]: one setState is enough is make changes change in state. Unnecessary use of setState

File: src/pages/InputDemo/InputDemo.jsx [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+    handlePositionChange = (e) => {
+      const { sport } = this.state;
+      return sport === 'cricket' ? this.setState({ cricket: e.target.value }, () => console.log(this.state)) : this.setState({ football: e.target.value }, () => console.log(this.state));
+    }
```
Comment (namanpar) [inline src/pages/InputDemo/InputDemo.jsx:34]: create separate functions to handle changes on both selection (cricket and football)

**Review by namanpar — CHANGES_REQUESTED** [review]
Kindly fix UI as per task and read task carefully

File: src/config/constants.js (hunk 1 above)
Comment (shashankbaranwal) [reply src/config/constants.js:6]: Renamed from selectOptions to sports.

File: src/config/constants.js (hunk 2 above)
Comment (shashankbaranwal) [reply src/config/constants.js:16]: Changed Named to cricketOptions.

File: src/config/constants.js (hunk 3 above)
Comment (shashankbaranwal) [reply src/config/constants.js:35]: renamed to footballOptions.

File: src/components/TextField/TextField.jsx [hunk 8]
```diff
@@ -3,26 +3,30 @@ import PropTypes from 'prop-types';
...
-    <Input type="text" value={value} disabled={disabled} />
+    <Input type="text" value={value} disabled={disabled} onChange={onChange} />
   );
 };
 export default TextField;
 TextField.propTypes = {
   value: PropTypes.string.isRequired,
   disabled: PropTypes.bool,
   error: PropTypes.string,
+  onChange: PropTypes.func,
```
[the remark this answers is not in the document]
Comment (shashankbaranwal) [reply src/components/TextField/TextField.jsx:26]: Changed is required prop

File: src/pages/InputDemo/InputDemo.jsx (hunk 7 above)
Comment (shashankbaranwal) [reply src/pages/InputDemo/InputDemo.jsx]: Created separate function to handle changes on both selection.

File: src/pages/InputDemo/InputDemo.jsx (hunk 6 above)
Comment (shashankbaranwal) [reply src/pages/InputDemo/InputDemo.jsx]: changed to one state.

File: src/pages/InputDemo/InputDemo.jsx (hunk 5 above)
Comment (shashankbaranwal) [reply src/pages/InputDemo/InputDemo.jsx]: remove functions, if-else and optimize code as per the requirement.

File: src/components/TextField/TextField.jsx [hunk 9]
```diff
@@ -3,26 +3,30 @@ import PropTypes from 'prop-types';
...
 };
 TextField.defaultProps = {
   disabled: false,
   error: '',
+  onChange: '',
```
[the remark this answers is not in the document]
Comment (shashankbaranwal) [reply src/components/TextField/TextField.jsx]: Changed proptypes function as a empty function.

File: src/pages/InputDemo/InputDemo.jsx (hunk 4 above)
Comment (shashankbaranwal) [reply src/pages/InputDemo/InputDemo.jsx]: Created a function without selection options.

**Pull request merged**